repair: optimise  repair reader with different shard count

gusev-p opened the pull request:
Consider a cluster with no data, e.g. in tests. When a new node is bootstrapped with repair we iterate over all (shard, table, range), read data from all the peer nodes for the range, look for any discrepancies and heal them. Even for small num_tokens (16 in the tests) the number of affected ranges (those we need to consider) amounts to total number of tokens in the cluster, which is 32 for the second node and 48 for the third. Multiplying this by the number of shards and the number of tables in each keyspace gives thousands of ranges. For each of them we need to follow some row level repair protocol, which includes several RPC exchanges between the peer nodes and creating some data structures on them. These exchanges are processed sequentially for each shard, there are `parallel_for_each` in code, but they are throttled by the choosen memory constraints and in fact execute sequentially.

When the bootstrapping node (master) reaches a peer node and asks for data in the specific range and master shard, two options exist. If sharder parameters (primarily, `--smp`) are the same on the master and on the peer, we can just read one local shard, this is fast. If, on the other hand, `--smp` is different, we need to do a multishard query. The given range from the master can contain data from different peer shards, so we split this range into a number of subranges such that each of them contain data only from the given master shard (`dht::selective_token_range_sharder`). The number of these subranges can be quite big (300 in the tests). For each of these subranges we do `fast_forward_to` on the `multishard_reader`, and this incurs a lot of overhead, mainly becuse of `smp::submit_to`.

In this series we optimize this case. Instead of splitting the master range and reading only what's needed, we read all the data in the range and then apply the filter by the master shard. We do this if the estimated number of partitions is small (<=100).

This is the logs of starting a second node with `--smp 4`, first node was `--smp 3`:

```
with this patch
    20:58:49.644 INFO> [debug/topology_custom.test_topology_smp.1] starting server at host 127.222.46.3 in scylla-2...
    20:59:22.713 INFO> [debug/topology_custom.test_topology_smp.1] started server at host 127.222.46.3 in scylla-2, pid 1132859

without this patch
    21:04:06.424 INFO> [debug/topology_custom.test_topology_smp.1] starting server at host 127.181.31.3 in scylla-2...
    21:06:01.287 INFO> [debug/topology_custom.test_topology_smp.1] started server at host 127.181.31.3 in scylla-2, pid 1134140
```

Fixes: #14093

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1711/

denesb commented on replica/database.cc:
```diff
@@ -2849,14 +2852,29 @@ flat_mutation_reader_v2 make_multishard_streaming_reader(distributed<replica::da
+        range,
+        full_slice);
+    return make_filtering_reader(std::move(multishard_reader), std::move(filter));
+}
```
I think it would be better if you selected the parameters and added the filtering on top in the caller.

denesb replied:
Or in other words, not add a new overload but just pick the params for the existing `make_...` method.

gusev-p replied:
You mean a `range_generator` returning one single range, which is the entire range from the repair master? It feels a bit awkward to me, and it leaves some overhead from `make_flat_multi_range_reader`. An overload with explicit `dht::partition_range` parameter is more straightforward. I can move filtering to the caller, though.

denesb replied:
Ah, I missed that we have a range generator parameter. Indeed that merits another overload, with the filtering stage left for the caller to do.

denesb commented on repair/row_level.cc:
```diff
@@ -979,7 +998,15 @@ class repair_meta {
+                    if (_repair_master || _same_sharding_config) {
+                        return repair_reader::read_strategy::local;
+                    }
+                    if (_estimated_partitions <= 100) {
```
Instead of this hard-coded limit of 100, it would be better to compare against something like the number of ranges. Note that discarding a partition also incurs a possible `smp::submit_to()` to each shards, because the filter calls `next_partition()`. So in essence, this optimization relies on getting away with fewer `next_partition()` then `fast_forward_to()`.

asias replied:
The row level repair always works on a single token range. The sharder will convert the single range into lots of subranges. The number of token ranges depends on the token range, shard number and msb. Assume a single token range is translated to N subranges.

With current split approach:
- Each sub reader of the multishard reader will work on N sub ranges. 
- No need to drop any partitions
 
With the filter approach:
- Each sub reader of the multishard reader will work on 1 original ranges. 
- We need to drop any partitions that does not needed

It is possible that the subrange we need contains few data but the unwanted subranges contains lots of data. It is a big waste to read those partitions. On the other hand, the with first approach, we worked on more sub ranges, but at least we do not waste disk io to read unwanted partitions and drop.  

It is hard to say which one is more efficient and select which one to use is better.

asias replied:
@gusev-p Do you see noticeable difference with release / dev build with your patch?

denesb replied:
@asias note that `fast_forward_to()` (used to skip between ranges) is not free either. It involves I/O (possibly) to read the index. Filtering also doesn't mean the entire partition is read (which could be a huge waste indeed). The reader will read a buffer worth of data from a partition, or if the partition is small, multiple partitions until the buffer is full. The filter will then instruct the reader to skip to the next partition. If the partition is small, and fits into the buffer entirely, the reader just has to discard data from the buffer. If the partition is large, only a buffer-full is thrown away, and the reader will do a skip using the index (possibly doing I/O).

So, neither variant is free and there can be cases where either will be cheaper than the other. The logic proposed in this patch *does* have the possibility to speed up repair. What I'm not confident is the heuristics it bases this choice upon.

gusev-p replied:
> @gusev-p Do you see noticeable difference with release / dev build with your patch?

```
with optimisation:
    16:52:21.951 INFO> [release/topology_custom.test_topology_smp.1] starting server at host 127.75.65.3 in scylla-2...
    16:52:24.608 INFO> [release/topology_custom.test_topology_smp.1] started server at host 127.75.65.3 in scylla-2, pid 1154689

    24608 - 21951 = 2657

without optimisation:
    17:01:25.408 INFO> [release/topology_custom.test_topology_smp.1] starting server at host 127.77.171.3 in scylla-2...
    17:01:28.987 INFO> [release/topology_custom.test_topology_smp.1] started server at host 127.77.171.3 in scylla-2, pid 1155934

    28987 - 25408 = 3579
```

This optimisation is mostly for debug builds.

denesb reviewed (changes requested):
I'm not completely sold on this heuristics. The estimates are exactly that, estimates and they can be way off. When we have no idea how many partitions some sstable will have, we just put `1` as the estimate. I'm afraid this can backfire and make repair much slower on some customer cluster.

That said I don't have a better idea on how to improve repair performance on mixed-smp cluster, save for a crazy shared mutishard reader idea, which even I don't like. This is a very real problem that not only tests, but many customers are suffering from and a solution eludes us.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1712/

asias commented:
> I'm not completely sold on this heuristics. The estimates are exactly that, estimates and they can be way off. When we have no idea how many partitions some sstable will have, we just put `1` as the estimate. I'm afraid this can backfire and make repair much slower on some customer cluster.
> 
> That said I don't have a better idea on how to improve repair performance on mixed-smp cluster, save for a crazy shared mutishard reader idea, which even I don't like. This is a very real problem that not only tests, but many customers are suffering from and a solution eludes us.

I want mention what this patch tries to solve is different from the reader eviction problems with different shard count we saw. This patch does not reduce the number of sub readers of the multi-shard reader but reduces the number of ranges each sub reader would work on at the expense of reading unwanted partitions and dropping it.

gusev-p commented:
> When we have no idea how many partitions some sstable will have, we just put 1 as the estimate.

I don't see that in code, could you point it to me? `sstable::estimated_keys_for_range` just extrapolates linearly pages to keys.

I see other strange thing with partition count estimation, [here](https://github.com/scylladb/scylladb/blob/master/repair/row_level.cc#L963) we `repeat` over all the subranges that `selective_token_range_sharder` gives to us, but we don't use the concrete subrange, we just call the same `do_estimate_partitions_on_all_shards` many times, which count partitions in the master range. Now we [actually don't use estimated_partition](https://github.com/scylladb/scylladb/issues/14093#issuecomment-1580661892) in any way, so this went unnoticed.

> The estimates are exactly that, estimates and they can be way off.

Maybe, but what `sstable::estimated_keys_for_range` can tell for sure is whether the range is empty or not. I mean, if this method returns 0 we can rely on this. We can use `multishard_split` only if estimated_partitions on this concrete peer node is zero, we don't need global estimated_partitions (sum for all nodes) for this. Even in this variant the optimization will work, since almost all of the ranges in my case are empty. Moreover, maybe in this case we can avoid creating a reader at all and immediately return an empty `sync_boundary`.

gusev-p commented:
> Moreover, maybe in this case we can avoid creating a reader at all and immediately return an empty sync_boundary.

Current implementation of `estimate_partitions` doesn't account for memtables, some data may reside only there, given up on this.

gusev-p commented:
v2:
  * get_estimated_partitions [fix](https://github.com/scylladb/scylladb/pull/14178/commits/c4c1af0c096023b8c364c0167584168e744dd612);
  * use local peer estimated_partitions instead of global sum;
  * changed hardcoded limit of 100 partitions to the heuristic proposed above: use filter if the number of partitions is smaller than the number of ranges

@denesb

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1741/

asias commented on repair/row_level.cc:
```diff
@@ -979,6 +998,42 @@ class repair_meta {
         return with_gate(_gate, [this, estimated_partitions] {
             _estimated_partitions = estimated_partitions;
             _repair_writer->set_estimated_partitions(_estimated_partitions);
+            _repair_reader.emplace(_db,
```
Creating the reader is not directly related to set_estimated_partitions. I think we have a better place to delay the creation of the reader. We can do that in read_rows_from_disk:
```
read_rows_from_disk() {
       if (!_repair_reader) {
              create the reader
       }
       ....
}
```
This guarantees we always create the reader before we read it.

asias commented on repair/row_level.cc:
```diff
@@ -979,6 +998,42 @@ class repair_meta {
+                    // master shard, we call next_partition(), which effectively clears
+                    // the buffer until the next partition is reached.
+                    if (_local_range_estimation &&
+                        _local_range_estimation->partitions_count <= _local_range_estimation->master_subranges_count)
```
Let's add a debug log about the partitions_count,  master_subranges_count and the read strategy selection. 

Perhaps we can relax the check a bit, e.g., 


_local_range_estimation->partitions_count * factor <= _local_range_estimation->master_subranges_count

factor = 0.5 or 0.8

gusev-p replied:
I added the debug logs.

I'm not quite into the `factor` thing though, it seems to blur the heuristic reasoning outlined in the comment.

asias replied:
How about adding the keyspace={}, table={}, range={}, as well. You can get it with: _schema->ks_name()  _schema->cf_name().

asias replied:
The factor thing is like saying if _local_range_estimation->master_subranges_count / _local_range_estimation->partitions_count >= factor, we apply the filter reader trick. 

It does not have to be 1. The benefits of the filter reader probably shows up with a smaller factor already. 

With avi's unit test suggestions, we can play with different configurations to find a better empirical factor.

gusev-p replied:
> How about adding the keyspace={}, table={}, range={}, as well. You can get it with: _schema->ks_name() _schema->cf_name().

ok. I noticed that most of the `debug/trace` logs in `row_level.cc` lacks this 'context-binding' information, such as these table/range or just `repair_meta_id`. It might make it difficult to reason about and draw correlations. Perhaps we should revisit that in separate PR.

gusev-p replied:
> The factor thing is like saying if _local_range_estimation->master_subranges_count / _local_range_estimation->partitions_count >= factor, we apply the filter reader trick.
> 
> It does not have to be 1. The benefits of the filter reader probably shows up with a smaller factor already.
> 
> With avi's unit test suggestions, we can play with different configurations to find a better empirical factor.

I looked at the actual values which `get_estimated_partitions` returns, in my test there were 300 partitions spread around two shards, master range is `open_ended_both_sides`, subranges=4000, estimated partitions=126298. This is because of the accumulated error, 
total_keys=300 total_pages=16 estimated_keys_for_subrange=16, total =16 *4000=64000, which is multiplied by two shards.

The point it, `get_estimated_partitions` doesn't work if the number of subranges is big, so currently there is no point in tuning the correlation with additional coefficients.

I don't have better idea for this heuristic. I left it as is, since it still solves the initial problem with slow repair in debug mode - `get_estimated_partitions` returns zero for most of the ranges and we use fast `multishard_filter`.

gusev-p commented:
v3: 
  * moved `_repair_reader` creation to `read_rows_from_disk`;
  * added debug logs to read_strategy selection logic;
  * splitted the `make_multishard_streaming_reader` commit.

@asias

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1773/

avikivity commented:
100 partitions can be a huge amount of data if the partitions are very large.

avikivity commented on repair/row_level.cc:
```diff
@@ -283,40 +298,55 @@ using is_local_reader = bool_class<class is_local_reader_tag>;
+                        return remote_sharder.shard_of(k.token()) == remote_shard;
+                    });
+            }
         }
```
This won't compile on gcc, which allows enums to have non-enumerated values when considering control flow. So it considers it legal of the switch to not match anything (it will warn like clang) and then error on a function with no return value.

The way to fix it is to add an on_internal_error() (which is marked [[noreturn]]).

gusev-p replied:
fixed

avikivity commented on repair/row_level.cc:
```diff
@@ -707,6 +737,11 @@ class repair_meta {
+    struct local_range_estimation {
+        size_t master_subranges_count;
+        size_t partitions_count;
+    };
```
Counting partitions is wrong, as partitions can be very large.

Better to count bytes, we have byte counts for sstables and memtables already.

We should consider the number of bytes that are going to be discarded. Say, if we're going to discard more than 1MB, then better to use shard hopping. I guess the number of bytes to be discarded =~ total number of bytes in the table, so it's enough to check if the table size is smaller than a megabyte.

gusev-p replied:
> I guess the number of bytes to be discarded =~ total number of bytes in the table, so it's enough to check if the table size is smaller than a megabyte.

We can have a small sized table with large number of partitions, `multishard_filter` will have a lot of overhead from `next_partition` calls in this case. The heuristic with `master_subranges_count ` and `partitions_count` tries to minimise the number of hops, either `next_partition` for  `multishard_filter` or `fast_forward_to` for `multishard_split`, leaving discarded bytes overhead aside. This is elaborated in [this comment](https://github.com/scylladb/scylladb/pull/14178/files#diff-e8909a2855e30487ffe474441ada239cfe8401fe46b51d0b28718fac529d82dbR1101).

avikivity commented:
> I'm not completely sold on this heuristics. The estimates are exactly that, estimates and they can be way off. When we have no idea how many partitions some sstable will have, we just put `1` as the estimate. I'm afraid this can backfire and make repair much slower on some customer cluster.
> 
> That said I don't have a better idea on how to improve repair performance on mixed-smp cluster, save for a crazy shared mutishard reader idea, which even I don't like. This is a very real problem that not only tests, but many customers are suffering from and a solution eludes us.

Tablets

gusev-p commented:
> 100 partitions can be a huge amount of data if the partitions are very large.

As @denesb pointed out above, the code doesn't imply we would read all the partitions data, only the limited buffer inside `flat_mutation_reader_v2` which will be discarded on `next_partition()`.

gusev-p commented:
> I'd like a dedicated unit test of the new multishard_reader mode.

@asias @denesb could you please advise how to do this efficiently?

avikivity commented:
> > 100 partitions can be a huge amount of data if the partitions are very large.
> 
> As @denesb pointed out above, the code doesn't imply we would read all the partitions data, only the limited buffer inside `flat_mutation_reader_v2` which will be discarded on `next_partition()`.

That's true.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1777/

asias commented:
> > I'd like a dedicated unit test of the new multishard_reader mode.
> 
> @asias @denesb could you please advise how to do this efficiently?

There is a test/boost/repair_test.cc which tests the repair writer. Try extending it to test repair_reader.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1943/

gusev-p commented:
v4:
  * added unit test for `repair_reader`

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1969/

gusev-p commented:
@asias please review

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1990/

gusev-p commented:
@asias ping

gusev-p commented:
@avikivity could you have another look at this? I added a unit test for `repair_reader`.

gusev-p commented:
@avikivity @asias megaping

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2235/

asias commented on test/boost/repair_test.cc:
```diff
@@ -133,3 +136,102 @@ SEASTAR_TEST_CASE(flush_repair_rows_on_wire_to_sstable) {
+        co_await do_check(dht::sharder(local_sharder.shard_count() + 1, local_sharder.sharding_ignore_msb()),
+            repair_reader::read_strategy::multishard_split,
+            repair_reader::read_strategy::multishard_filter);
+        co_await do_check(local_sharder,
```
How about adding one more check for local  vs split

```    
    co_await do_check(local_sharder,
            repair_reader::read_strategy::local,
            repair_reader::read_strategy::multishard_split);
```

asias replied:
local vs split
local vs filter 

are the two we use in practice.

gusev-p replied:
added, please approve

asias commented:
I have a minor comment on the unit test. Otherwise looks good tome. Thanks @gusev-p !

asias commented:
Btw, you can set the title like: `repair: optimise  repair reader with different shard count`.  Row level repair is the only repair we support now.

avikivity commented:
I'm fine with the series, will await @asias confirmation.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2270/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2277/

asias commented:
> I'm fine with the series, will await @asias confirmation.

Looks good. Ping for merge. Thanks.

gusev-p commented:
@avikivity another megaping

kbr-scylla commented:
> @avikivity another megaping

you should try PINGUS MAXIMUS

yaronkaikov commented:
@avikivity please merge :+1:

avikivity commented:
Pings with reused references are penalized. Exercise creativity.
